### PR TITLE
Refactors JSON schema and adds SHA256 verification of tarballs

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This change allows us increased confidence that tarballs downloaded were
what we expected. Notably, `getenvoy use N.N.N` will fail now, if the
`sha256sum` doesn't match what's in the json key for that filename.

This refactors out Envoy-specific terminology from the JSON schema used
to validate envoy-versions.json. It is now called
release-versions-schema.json.

This adds a new key sha256sums in the schema and backfills data for the
existing versions.

The key name "sha256sums" comes from the `sha256sum` command. This is
SHA-256 instead of SHA-512 as it is a ubuquitous hash algo, for example,
used by default also in goreleaser.

This doesn't use a more generic term like "hash" because doing so would
cause unnecessary confusion on what hash it is (ex MD-5?).

The map "sha256sums" allows for easy indexing, and using the filename
instead of the URL allows less changes when rehosting the exact same
files.

Signed-off-by: Adrian Cole <adrian@tetrate.io>

### CODE REVIEW

File: site/usage.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ choose one, invoke `getenvoy use 1.18.3`. This installs into
 
 You may want to override `$ENVOY_VERSIONS_URL` to supply custom builds or
 otherwise control the source of Envoy binaries. When overriding, validate
-your JSON first: https://getenvoy.io/envoy-versions-schema.json
+your JSON first: https://getenvoy.io/release-versions-schema.json
```
Comment (codefromthecrypt) [inline site/usage.md:11]: I will add a redirect for this post review and pre-merge

File: internal/envoy/install.go [hunk 2]
```diff
@@ -84,19 +94,42 @@ func verifyEnvoy(installPath string) (string, error) {
...
 
-func untarEnvoy(ctx context.Context, dst, url, p, v string) error { // dst, src order like io.Copy
-	resp, err := httpGet(ctx, url, p, v)
+type digester struct {
+	r io.Reader
+	h hash.Hash
+	e error
+}
+
+func (d *digester) Read(p []byte) (n int, err error) {
```
Comment (codefromthecrypt) [inline internal/envoy/install.go:103]: I couldn't find another non-buffering way to do an incremental digest

File: internal/version/version.go [hunk 3]
```diff
@@ -12,40 +12,62 @@
...
 package version
 
 import _ "embed" // We embed the Envoy version so that we can cache it in CI
 
+var getEnvoy = "dev"
+
 // GetEnvoy is the version of the CLI, used in help statements and HTTP requests via "User-Agent".
-// Override this via "-X github.com/tetratelabs/getenvoy/internal/version.GetEnvoy=XXX"
-var GetEnvoy = "dev"
+// Override this via "-X github.com/tetratelabs/getenvoy/internal/version.getEnvoy=XXX"
```
Comment (codefromthecrypt) [inline internal/version/version.go:23]: I checked this works

File: internal/version/version.go [hunk 4]
```diff
@@ -12,40 +12,62 @@
...
+
+// SHA256Sum is a SHA-256 lower-hex hash. Ex. "1274f55b3022bc1331aed41089f189094e00729981fe132ce00aac6272ea0770"
+type SHA256Sum string
+
+// ReleaseDate is the publish date of the release Version. Ex. "2021-05-11"
+type ReleaseDate string
+
+// Release primarily maps available Tarballs for a Version
+type Release struct {
+	ReleaseDate
```
Comment (mathetake) [inline internal/version/version.go:69]: IMHO we should avoid embedding structs as much as possible unless it's necessary (for example, in the case where we want to provide default implementation of interfaces via embedding).

Comment (mathetake) [reply internal/version/version.go:69]: often this make it difficult at call sites.

File: internal/version/version.go [hunk 5]
```diff
@@ -12,40 +12,62 @@
...
+type Tarball string
+
+// TarballURL is the HTTPS URL to the Tarball. SHA256Sums must include its base name.
+type TarballURL string
+
+// SHA256Sum is a SHA-256 lower-hex hash. Ex. "1274f55b3022bc1331aed41089f189094e00729981fe132ce00aac6272ea0770"
+type SHA256Sum string
+
+// ReleaseDate is the publish date of the release Version. Ex. "2021-05-11"
+type ReleaseDate string
```
Comment (mathetake) [inline internal/version/version.go:65]: I'm skeptical that we need these aliasing.. I see a lot of `string(...)` anywhere.

File: internal/cmd/versions.go [hunk 6]
```diff
@@ -80,10 +80,8 @@ func NewVersionsCmd(o *globals.GlobalOpts) *cli.Command {
 }
 
 type versionReleaseDate struct {
-	// version ex "1.15.5"
-	version string
-	// releaseDate ex "2021-05-11"
-	releaseDate string
+	version.Version
+	version.ReleaseDate
```
Comment (mathetake) [inline internal/cmd/versions.go:84]: same here for embedding and aliasing.

File: internal/version/version.go (hunk 5 above)
Comment (mathetake) [reply internal/version/version.go:65]: my point is I think having descriptive field or variable names is enough vs bother doing type conversion a lot

**Review by mathetake — COMMENTED** [review]
other than the comment above, the change LGTM!

File: internal/envoy/install.go [hunk 7]
```diff
@@ -48,16 +52,22 @@ func InstallIfNeeded(ctx context.Context, o *globals.GlobalOpts, p, v string) (s
 			return "", fmt.Errorf("couldn't find version %q for platform %q", v, p)
 		}
 
+		tarball := version.Tarball(path.Base(string(tarballURL)))
+		sha256Sum := ev.SHA256Sums[tarball]
+		if sha256Sum == "" {
```
Comment (mathetake) [inline internal/envoy/install.go:57]: more precisely I suggest check `len(sha256Sum) == 64` 😄

File: internal/version/version.go (hunk 5 above)
Comment (codefromthecrypt) [reply internal/version/version.go:65]: I kindof agree, though I would say it is pretty limited and this found a bug or two.

Comment (codefromthecrypt) [reply internal/version/version.go:65]: the maintenance of this vs json schema is far less hairy with type aliases

File: internal/version/version.go (hunk 4 above)
Comment (codefromthecrypt) [reply internal/version/version.go:69]: that's fine I just noticed it was less typing and work

File: internal/version/version.go (hunk 5 above)
Comment (codefromthecrypt) [reply internal/version/version.go:65]: the other thing is the these are coordinated map keys. the aliases do help on that part

Comment (mathetake) [reply internal/version/version.go:65]: OK fair enough! better leave them as-is

File: site/envoy-versions-schema.json [hunk 8]
```diff
@@ -1,96 +0,0 @@
-{
```
Comment (mathetake) [inline site/envoy-versions-schema.json:1]: is this going to be a breaking change?

Comment (codefromthecrypt) [reply site/envoy-versions-schema.json:1]: we don't have anything except docs referring to the location of this file. but I will rename it back for this PR anyway!

File: internal/version/version.go [hunk 9]
```diff
@@ -12,40 +12,62 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// Package version declares types for each string to keep strict coupling to the JSON schema
```
Comment (codefromthecrypt) [inline internal/version/version.go:15]: this is a comment which can be used to squarely blame adrian for the type aliases later. If they become a problem I won't complain much about reverting them.. I also usually find type aliases annoying

File: site/envoy-versions-schema.json [hunk 10]
```diff
@@ -73,7 +98,7 @@
       },
       "required": [
         "releaseDate",
-        "downloads"
+        "tarballs"
```
Comment (codefromthecrypt) [inline site/envoy-versions-schema.json:101]: typo!

File: site/envoy-versions-schema.json (hunk 8 above)
Comment (codefromthecrypt) [reply site/envoy-versions-schema.json:1]: ps to answer no.. it isn't a breaking change, but before our json schema had a bug in it, so there's that also :D

**Comment by codefromthecrypt** [comment]
3Q for the help @mathetake!

**Pull request merged**